File: Jan-2023/1.java

//290. Word Pattern

class Solution {
	public boolean wordPattern(String pattern, String s) {
		//Using a HashMap for mapping each word to a character in pattern
		HashMap<String, Character> map = new HashMap<>();
		//Converting the string of words into an array 
		String[] array = s.split(" ");
		if (array.length != pattern.length()) return false;
		for (int i = 0; i < pattern.length(); i++) {
			if (map.containsKey(array[i])) {
				//comparing if the character in pattern is same as the value in the HashMap
				if (pattern.charAt(i) != map.get(array[i])) {
					System.out.println(array[i]);
					return false;
				}
			}
			else {
				if (map.containsValue(pattern.charAt(i))) return false;
				map.put(array[i], pattern.charAt(i));
			}
		}
		return true;
	}
}

File: Jan-2023/C++/1.cpp
class Solution
{
  public:
	bool wordPattern(string pattern, string s)
	{
                //Using a HashMap for mapping each word to a character in pattern
		unordered_map<string, char> map;
		vector<string> array;
		string temp;
                //Converting the string of words into an vector array
		for (int i = 0; i < s.size(); i++)
		{
			if (s[i] != ' ')
			{
				temp += s[i];
			}
			else
			{
				array.push_back(temp);
				temp = "";
			}
			if (i == s.size() - 1)
				array.push_back(temp);
		}

		if (pattern.size() != array.size())
			return false;

		for (int i = 0; i < pattern.size(); i++)
		{
			if (map.count(array[i]))
			{
                                //comparing if the character in pattern is same as the value in the HashMap
				if (pattern[i] != map.at(array[i]))
				{
					return false;
				}
			}
			else
			{
				for (const auto &iter : map)
				{
					if (iter.second == pattern[i])
					{
						return false;
					}
				}
				map[array[i]] = pattern[i];
			}
		}
		return true;
	}
};
